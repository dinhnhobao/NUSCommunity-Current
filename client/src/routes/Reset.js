import React from "react";
import { Link } from "react-router-dom";
import Router from 'next/router'
import { Form, Input, Button, Row, Col } from "antd";
import icon from "../assets/favicon.ico";
import "../css/Reset.css";
import Session from '../utils/session';


  // const onFinish = (values) => {
  //   console.log(values);
  //   alert("A verification link has been sent to your email");
  //   window.location.href = "/login";
  // };
class Reset extends React.Component {
  static async getInitialProps({ req, res }) {

    let props = {
        session: ''
    }

    if (req && req.session) {
        props.session = req.session
    } else {
        props.session = await Session.getSession()
    }

    if (props.session && props.session.loggedin) {
        if (req) {
            res.redirect('/')
        } else {
            Router.push('/')
        }
    }

    return props
  }
  constructor(props) {
    super(props)
    this.state = {
        email: '',
        password: '',
        confirmPassword: '',
        message: null
    }
    this.handleEmailChange = this.handleEmailChange.bind(this)
    this.handlePasswordChange = this.handlePasswordChange.bind(this)
    this.handleConfirmPasswordChange = this.handleConfirmPasswordChange.bind(this)
    this.handleReset = this.handleReset.bind(this)
}

  handleEmailChange(event) {
    this.setState({
        email: event.target.value.trim()
    })
  }

  handlePasswordChange(event) {
    this.setState({
        password: event.target.value.trim()
    })
  }

  handleConfirmPasswordChange(event) {
    this.setState({
        confirmPassword: event.target.value.trim()
    })
  }

  handleReset(event) {
    event.preventDefault()
    
    this.setState({
        message: null
    })

    if (!this.state.email || !this.state.password || !this.state.confirmPassword) {
        this.setState({
            message: 'All fields are required!'
        })

        return
    }

    if (this.state.password !== this.state.confirmPassword) {
        this.setState({
            message: 'Password did not match!'
        })

        return
    }

    let data = {
        email: this.state.email,
        password: this.state.password,
    }

    console.log(this.state.email + " " + this.state.password);

    fetch('auth/reset', {
        method: 'POST',
        body: JSON.stringify(data),
        headers: {
            'Content-Type': 'application/json'
        }
    })
    .then(res => res.json())
    .then(response => {
        if (response.message) {
            this.setState({
                message: response.message
            })
        } else if (response.email) {
            this.setState({
              message: "A verification link has been sent to " + response.email
            })
            //alert("A verification link has been sent to " + response.email);
            //Router.push('/check-email?email=' + response.email)
        } else {
            this.setState({
                message: 'Unknown Error!'
            })
        }
    })
    .catch(error => {
        console.error('Error:', error)
        this.setState({
            message: 'Request Failed!'
        })
    })
}

  
  render() {
    const alert = (this.state.message === null) ? <div/> : <div className="alert-style" role="alert">{this.state.message}</div>
    return (
    <div style={{ paddingTop: 36 }}>
      <div className="reset">
        <Row gutter={16}>
          <Col span={6}>
            <div className="reset-left">
              <img className="reset-left-img" src={icon} alt="" />
              <Link className="reset-left-btn" to="/login">
                Login
              </Link>
            </div>
          </Col>
          <Col span={18}>
            <div className="reset-right">
              <div className="reset-right-title">Reset Password Form</div>
              <Form
                name="normal_reset"
                className="reset-right-form"
              >
                <Form.Item
                  name="email"
                  rules={[
                    {
                      required: true,
                      message: "Please enter a valid NUS email!",
                    },
                  ]}
                >
                  <Input placeholder="NUS email" name="email" addonAfter="@u.nus.edu" value={this.state.email} onChange={this.handleEmailChange}/>
                </Form.Item>
                <Form.Item
                  name="password"
                  rules={[
                    {
                      required: true,
                      message: "Please input your password!",
                    },
                  ]}
                >
                  <Input type="password" name="password" placeholder="New Password" value={this.state.password} onChange={this.handlePasswordChange}/>
                </Form.Item>
                <Form.Item
                  name="confirm"
                  dependencies={["password"]}
                  rules={[
                    {
                      required: true,
                      message: "Please input your password again！",
                    },
                    ({ getFieldValue }) => ({
                      validator(rule, value) {
                        if (!value || getFieldValue("password") === value) {
                          return Promise.resolve();
                        }
                        // eslint-disable-next-line prefer-promise-reject-errors
                        return Promise.reject(
                          "The confirm password does not match with the password!"
                        );
                      },
                    }),
                  ]}
                >
                  <Input type="password" name="confirmPassword" placeholder="Confirm New Password" value={this.state.confirmPassword} onChange={this.handleConfirmPasswordChange} />
                </Form.Item>
                {alert}
                <Form.Item>
                  <Button
                    type="submit"
                    onClick={this.handleReset}
                    className="reset-right-form-button"
                    size="large"
                  >
                    Reset
                  </Button>
                </Form.Item>
              </Form>
            </div>
          </Col>
        </Row>
      </div>
    </div>
    )};
}

export default Reset;
