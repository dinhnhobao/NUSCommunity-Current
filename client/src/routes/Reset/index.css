.reset {
  background: #35d0ba;
  width: 1140px;
  margin: 0 auto;
  padding: 36px;
}
.reset-left {
  margin-top: 48px;
}
.reset-left-img {
  margin: 0 auto;
  margin-top: 70px;
  margin-bottom: 10px;
  width: 200px;
  display: block;
}
.reset-left-btn {
  border-radius: 24px;
  padding: 5px;
  width: 140px;
  background: #f8f9fa;
  font-weight: bold;
  color: #383d41;
  margin: 0 auto;
  margin-top: 70px;
  margin-bottom: 8px;
  cursor: pointer;
  text-align: center;
  display: block;
  font-weight: bolder;
}
.reset-right {
  padding-top: 80px;
  width: 100%;
  background: #f8f9fa;
  border-top-left-radius: 10% 50%;
  border-bottom-left-radius: 10% 50%;
}
.reset-right-title {
  text-align: center;
  font-weight: bolder;
  margin-bottom: 46px;
  line-height: 34px;
  font-size: 28px;
  color: #495057;
}
.reset-right-form {
  margin: 0 100px;
}
.reset-right-form-button {
  display: block;
  width: 400px;
  margin: 0 auto;
  margin-top: 20px;
  margin-bottom: 80px;
}
.alert-style {
  text-align: center;
  color: red;
  font-weight: bold;
}