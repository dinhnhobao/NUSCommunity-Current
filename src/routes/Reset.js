import React from "react";
import { Link } from "react-router-dom";
import { Form, Input, Button, message, Row, Col } from "antd";
import icon from "../assets/favicon.ico";
import "../css/Reset.css";

const Reset = () => {
  const onFinish = (values) => {
    console.log(values);
    alert("A verification link has been sent to your email");
    window.location.href = "/login";
  };

  return (
    <div style={{ paddingTop: 36 }}>
      <div className="reset">
        <Row gutter={16}>
          <Col span={6}>
            <div className="reset-left">
              <img className="reset-left-img" src={icon} alt="" />
              <Link className="reset-left-btn" to="/login">
                Login
              </Link>
            </div>
          </Col>
          <Col span={18}>
            <div className="reset-right">
              <div className="reset-right-title">Reset Password Form</div>
              <Form
                name="normal_reset"
                className="reset-right-form"
                onFinish={onFinish}
              >
                <Form.Item
                  name="email"
                  rules={[
                    {
                      required: true,
                      message: "Please enter a valid NUS email!",
                    },
                  ]}
                >
                  <Input placeholder="NUS email" addonAfter="@u.nus.edu" />
                </Form.Item>
                <Form.Item
                  name="password"
                  rules={[
                    {
                      required: true,
                      message: "Please input your password!",
                    },
                  ]}
                >
                  <Input type="password" placeholder="New Password" />
                </Form.Item>
                <Form.Item
                  name="confirm"
                  dependencies={["password"]}
                  rules={[
                    {
                      required: true,
                      message: "Please input your password again！",
                    },
                    ({ getFieldValue }) => ({
                      validator(rule, value) {
                        if (!value || getFieldValue("password") === value) {
                          return Promise.resolve();
                        }
                        // eslint-disable-next-line prefer-promise-reject-errors
                        return Promise.reject(
                          "The confirm password does not match with the password!"
                        );
                      },
                    }),
                  ]}
                >
                  <Input type="password" placeholder="Confirm New Password" />
                </Form.Item>
                <Form.Item>
                  <Button
                    htmlType="submit"
                    type="primary"
                    className="reset-right-form-button"
                    ghost
                    size="large"
                  >
                    Reset
                  </Button>
                </Form.Item>
              </Form>
            </div>
          </Col>
        </Row>
      </div>
    </div>
  );
};

export default Reset;
