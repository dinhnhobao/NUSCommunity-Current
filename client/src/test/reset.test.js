import React from 'react';
import {shallow} from 'enzyme';
import Reset from '../routes/Reset/index';

const setUp = (props={}) => {
    const component = shallow(<Reset {...props} />);
    return component;
}

const findByTestAttr = (component, attr) => {
    const wrapper = component.find(`[data-test='${attr}']`);
    return wrapper;
}

describe('Reset Component', () => {
    let component;
    beforeEach(() => {
        component = setUp();
    })

    it('Should render correctly', () => {
        const wrapper =  findByTestAttr(component, 'resetComponent');
        expect(wrapper).toMatchSnapshot();
    });

    it('Should render without errors', () => {
        const wrapper =  findByTestAttr(component, 'resetComponent');
        expect(wrapper.length).toEqual(1);       
    });

    describe('Form', () => {
        it('Fill out email in the form', () => {
            const wrapper = shallow(<Reset/>);
            let emailInput = wrapper.find('Input').first();
            emailInput.simulate('change', { 
                target: {value: 'e0407663'},
            })
            emailInput = wrapper.find('Input').first();
            expect(emailInput.props().value).toEqual('e0407663')
        })

        it('Fill out password in the form', () => {
            const wrapper = shallow(<Reset/>);
            let passwordInput = wrapper.find('Input').last();
            passwordInput.simulate('change', { 
                target: {value: 'Anonymous'},
            })
            passwordInput = wrapper.find('Input').last();
            expect(passwordInput.props().value).toEqual('Anonymous')
        })
    })

    describe('Button', () => {
        it('Should render Button with click handle', () => {
            const wrapper =  findByTestAttr(component, 'buttonComponent');
            expect(wrapper.length).toEqual(1);
        })
    })
})
