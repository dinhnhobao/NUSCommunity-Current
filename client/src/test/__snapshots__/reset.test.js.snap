// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reset Component Should render correctly 1`] = `ShallowWrapper {}`;
